File: src/NetWasm.Hosting/Capabilities/CapabilityBindingValidator.cs
using System;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.Linq;
using NetWasm.Hosting.Deployment;
using NetWasm.Hosting.Execution;

namespace NetWasm.Hosting.Capabilities;

/// <summary>Proves that selected providers exactly satisfy and are authorized for a final deployment.</summary>
public sealed class CapabilityBindingValidator : ICapabilityBindingValidator
{
    private readonly IDeploymentManifestValidator _manifestValidator;
    private readonly INetWasmExecutionRequestValidator _requestValidator;
    private readonly IInternalImportPolicy _internalImports;

    public CapabilityBindingValidator(
        IDeploymentManifestValidator manifestValidator,
        INetWasmExecutionRequestValidator requestValidator)
        : this(manifestValidator, requestValidator, new InternalImportPolicy())
    {
    }

    internal CapabilityBindingValidator(
        IDeploymentManifestValidator manifestValidator,
        INetWasmExecutionRequestValidator requestValidator,
        IInternalImportPolicy internalImports)
    {
        _manifestValidator = manifestValidator ?? throw new ArgumentNullException(nameof(manifestValidator));
        _requestValidator = requestValidator ?? throw new ArgumentNullException(nameof(requestValidator));
        _internalImports = internalImports ?? throw new ArgumentNullException(nameof(internalImports));
    }

    public void Validate(
        DeploymentManifest manifest,
        NetWasmExecutionRequest request,
        ImmutableArray<NetWasmPlatformImportProvider> platformProviders,
        ImmutableArray<NetWasmApplicationImportProvider> applicationProviders)
    {
        ArgumentNullException.ThrowIfNull(manifest);
        ArgumentNullException.ThrowIfNull(request);
        _manifestValidator.Validate(manifest);
        _requestValidator.Validate(request);
        if (!string.Equals(manifest.BuildFingerprint, request.BuildFingerprint, StringComparison.Ordinal))
        {
            throw new ArgumentException("The execution request does not target this deployment build.", nameof(request));
        }

        if (platformProviders.IsDefault || applicationProviders.IsDefault)
        {
            throw new ArgumentException("Selected import provider collections must be explicit.", nameof(platformProviders));
        }

        var requiredModules = manifest.RequiredImportModules
            .Where(module => !_internalImports.IsInternal(module))
            .ToHashSet(StringComparer.Ordinal);
        var providerModules = new HashSet<string>(StringComparer.Ordinal);
        var platforms = ValidatePlatformProviders(platformProviders, requiredModules, providerModules, request.Grants);
        var applications = ValidateApplicationProviders(applicationProviders, requiredModules, providerModules);
        ValidateApplicationBindings(manifest, request, requiredModules, applications);
        ValidateRequiredImports(manifest.RequiredImports, platforms, applications);
        if (!providerModules.SetEquals(requiredModules))
        {
            throw new ArgumentException(
                "Every required import module must have exactly one selected provider.",
                nameof(platformProviders));
        }
    }

    private static Dictionary<string, NetWasmPlatformImportProvider> ValidatePlatformProviders(
        ImmutableArray<NetWasmPlatformImportProvider> providers,
        HashSet<string> requiredModules,
        HashSet<string> providerModules,
        NetWasmCapabilityGrants grants)
    {
        var validated = new Dictionary<string, NetWasmPlatformImportProvider>(StringComparer.Ordinal);
        foreach (var provider in providers)
        {
            ArgumentNullException.ThrowIfNull(provider);
            ValidateVersionedIdentifier(provider.Module);
            if (!IsReservedModule(provider.Module))
            {
                throw new ArgumentException("Platform providers must use reserved module keys.", nameof(providers));
            }

            ValidateProviderFunctions(provider.Module, provider.Functions);
            if (!requiredModules.Contains(provider.Module))
            {
                throw new ArgumentException("A selected platform provider is not required by the deployment.", nameof(providers));
            }

            if (!providerModules.Add(provider.Module))
            {
                throw new ArgumentException("Selected provider modules must be unique.", nameof(providers));
            }

            if (!IsGranted(provider.Capability, grants))
            {
                throw new ArgumentException("A required platform capability was not granted.", nameof(providers));
            }

            validated.Add(provider.Module, provider);
        }

        return validated;
    }

    private static Dictionary<string, NetWasmApplicationImportProvider> ValidateApplicationProviders(
        ImmutableArray<NetWasmApplicationImportProvider> providers,
        HashSet<string> requiredModules,
        HashSet<string> providerModules)
    {
        var validated = new Dictionary<string, NetWasmApplicationImportProvider>(StringComparer.Ordinal);
        foreach (var provider in providers)
        {
            ArgumentNullException.ThrowIfNull(provider);
            ValidateVersionedIdentifier(provider.Module);
            if (IsReservedModule(provider.Module))
            {
                throw new ArgumentException("Applications cannot provide reserved platform modules.", nameof(providers));
            }

            ValidateProviderFunctions(provider.Module, provider.Functions);
            if (!requiredModules.Contains(provider.Module))
            {
                throw new ArgumentException("A selected application provider is not required by the deployment.", nameof(providers));
            }

            if (!providerModules.Add(provider.Module))
            {
                throw new ArgumentException("Selected provider modules must be unique.", nameof(providers));
            }

            validated.Add(provider.Module, provider);
        }

        return validated;
    }

    private static void ValidateApplicationBindings(
        DeploymentManifest manifest,
        NetWasmExecutionRequest request,
        HashSet<string> requiredModules,
        Dictionary<string, NetWasmApplicationImportProvider> providers)
    {
        var artifacts = manifest.Artifacts.ToDictionary(artifact => artifact.RelativePath, StringComparer.Ordinal);
        var bindings = request.ApplicationImports.ToDictionary(binding => binding.Module, StringComparer.Ordinal);
        foreach (var binding in request.ApplicationImports)
        {
            if (!requiredModules.Contains(binding.Module))
            {
                throw new ArgumentException("An application import binding is not required by the deployment.", nameof(request));
            }

            if (!providers.ContainsKey(binding.Module))
            {
                throw new ArgumentException("An application import binding has no selected provider.", nameof(request));
            }

            if (!artifacts.TryGetValue(binding.ArtifactPath, out var artifact))
            {
                throw new ArgumentException("An application import artifact is missing from the deployment.", nameof(request));
            }

            if (!string.Equals(artifact.Role, "application-import", StringComparison.Ordinal)
                || !string.Equals(artifact.MediaType, "text/javascript", StringComparison.Ordinal)
                || !string.Equals(artifact.Sha256, binding.Sha256, StringComparison.Ordinal))
            {
                throw new ArgumentException("An application import does not match its deployment artifact contract.", nameof(request));
            }
        }

        foreach (var provider in providers.Values)
        {
            if (!bindings.ContainsKey(provider.Module))
            {
                throw new ArgumentException("A selected application provider has no request binding.", nameof(request));
            }
        }
    }

    private void ValidateRequiredImports(
        ImmutableArray<DeploymentFunction> requiredImports,
        Dictionary<string, NetWasmPlatformImportProvider> platformProviders,
        Dictionary<string, NetWasmApplicationImportProvider> applicationProviders)
    {
        foreach (var requiredImport in requiredImports)
        {
            if (_internalImports.IsInternal(requiredImport.Interface))
            {
                continue;
            }
            var functions = IsReservedModule(requiredImport.Interface)
                ? GetRequiredProvider(platformProviders, requiredImport.Interface).Functions
                : GetRequiredProvider(applicationProviders, requiredImport.Interface).Functions;
            var providedFunction = functions.SingleOrDefault(function =>
                string.Equals(function.Name, requiredImport.Name, StringComparison.Ordinal));
            if (providedFunction is null)
            {
                throw new ArgumentException("A selected provider is missing a required import member.", nameof(requiredImports));
            }

            if (!providedFunction.Parameters.SequenceEqual(requiredImport.Parameters, StringComparer.Ordinal)
                || !providedFunction.Results.SequenceEqual(requiredImport.Results, StringComparer.Ordinal))
            {
                throw new ArgumentException("A selected provider has an incompatible required import signature.", nameof(requiredImports));
            }
        }
    }

    private static TProvider GetRequiredProvider<TProvider>(
        Dictionary<string, TProvider> providers,
        string module)
        where TProvider : class
    {
        if (!providers.TryGetValue(module, out var provider))
        {
            throw new ArgumentException("A final required import has no selected provider.", nameof(providers));
        }

        return provider;
    }

    private static void ValidateProviderFunctions(string module, ImmutableArray<DeploymentFunction> functions)
    {
        if (functions.IsDefault)
        {
            throw new ArgumentException("Selected provider functions must be explicit.", nameof(functions));
        }

        var names = new HashSet<string>(StringComparer.Ordinal);
        foreach (var function in functions)
        {
            ArgumentNullException.ThrowIfNull(function);
            if (!string.Equals(function.Interface, module, StringComparison.Ordinal))
            {
                throw new ArgumentException("Provider functions must belong to their provider module.", nameof(functions));
            }

            ValidateFunctionName(function.Name);
            ValidateSignature(function.Parameters);
            ValidateSignature(function.Results);
            if (!names.Add(function.Name))
            {
                throw new ArgumentException("Provider function names must be unique within a module.", nameof(functions));
            }
        }
    }

    private static bool IsGranted(NetWasmPlatformCapability capability, NetWasmCapabilityGrants grants) => capability switch
    {
        NetWasmPlatformCapability.Baseline => true,
        NetWasmPlatformCapability.Environment => true,
        NetWasmPlatformCapability.PreopenedDirectories => true,
        NetWasmPlatformCapability.Network => grants.Network == NetWasmNetworkPolicy.AllowAll,
        NetWasmPlatformCapability.WallClock => grants.Clocks.Contains(NetWasmClock.Wall),
        NetWasmPlatformCapability.MonotonicClock => grants.Clocks.Contains(NetWasmClock.Monotonic),
        NetWasmPlatformCapability.Randomness => grants.Randomness,
        _ => throw new ArgumentOutOfRangeException(nameof(capability)),
    };

    private static bool IsReservedModule(string module) =>
        module.StartsWith("wasi:", StringComparison.Ordinal)
        || module.StartsWith("netwasm:platform/", StringComparison.Ordinal);

    private static void ValidateVersionedIdentifier(string value)
    {
        ValidateText(value);
        var separator = value.LastIndexOf('@');
        if (separator <= 0 || separator == value.Length - 1 || value.AsSpan(0, separator).Contains('@'))
        {
            throw new ArgumentException("Provider modules require one exact version.", nameof(value));
        }

        if (!(char.IsAsciiLetterLower(value[0]) || char.IsAsciiDigit(value[0])))
        {
            throw new ArgumentException("Provider modules must begin with a lowercase letter or digit.", nameof(value));
        }

        foreach (var character in value.AsSpan(0, separator))
        {
            if (!(char.IsAsciiLetterLower(character) || char.IsAsciiDigit(character) || character is '-' or '.' or ':' or '/'))
            {
                throw new ArgumentException("Provider modules must use canonical lowercase syntax.", nameof(value));
            }
        }

        var versionParts = value[(separator + 1)..].Split('.');
        if (versionParts.Length != 3 || versionParts.Any(part => part.Length == 0 || !part.All(char.IsAsciiDigit)))
        {
            throw new ArgumentException("Provider modules require an exact numeric version.", nameof(value));
        }
    }

    private static void ValidateFunctionName(string value)
    {
        if (!DeploymentFunctionName.IsCanonical(value))
        {
            throw new ArgumentException(
                "Provider function names must use canonical WIT identity syntax.",
                nameof(value));
        }
    }

    private static void ValidateSignature(ImmutableArray<string> values)
    {
        if (values.IsDefault)
        {
            throw new ArgumentException("Provider function signatures must be explicit.", nameof(values));
        }

        foreach (var value in values)
        {
            ValidateText(value);
        }
    }

    private static void ValidateText(string value)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(value);
        if (value.AsSpan().Trim().Length != value.Length || value.Any(char.IsControl))
        {
            throw new ArgumentException("Provider contract text must be canonical and contain no control characters.", nameof(value));
        }
    }
}

File: src/NetWasm.Hosting/JavaScript/capability-binding-validator.mjs
const factoryKeys = [
  "isInternalImport",
  "selectProviderKind",
  "validateManifest",
  "validateProviderMetadata",
  "validateRequest",
];
const requestKeys = ["applicationProviders", "manifest", "platformProviders", "request"];

export function createCapabilityBindingValidator(options) {
  assertExactDataObject(options, factoryKeys, "capability binding validator options");
  const {
    isInternalImport,
    selectProviderKind,
    validateManifest,
    validateProviderMetadata,
    validateRequest,
  } = options;
  if (typeof isInternalImport !== "function"
      || typeof selectProviderKind !== "function"
      || typeof validateManifest !== "function"
      || typeof validateProviderMetadata !== "function"
      || typeof validateRequest !== "function") {
    throw new TypeError("internal-import policy, provider-kind selector, and validation actions are required");
  }

  return Object.freeze(function validateCapabilityBinding(value) {
    assertExactDataObject(value, requestKeys, "capability binding request");
    const manifest = validateManifest(value.manifest);
    const request = validateRequest(value.request);
    if (manifest.buildFingerprint !== request.buildFingerprint) {
      throw new TypeError("execution request does not target the deployment build");
    }
    if (!Array.isArray(value.platformProviders)
        || !Array.isArray(value.applicationProviders)) {
      throw new TypeError("selected provider collections must be explicit");
    }

    const requiredModules = new Set(manifest.requiredImportModules.filter(
      module => !readInternalImport(isInternalImport, module)));
    const providerModules = new Set();
    const platformProviders = validatePlatformProviders(
      value.platformProviders,
      requiredModules,
      providerModules,
      request.grants,
      selectProviderKind,
      validateProviderMetadata);
    const applicationProviders = validateApplicationProviders(
      value.applicationProviders,
      requiredModules,
      providerModules,
      selectProviderKind,
      validateProviderMetadata);
    validateApplicationBindings(manifest, request, requiredModules, applicationProviders);
    validateRequiredImports(
      manifest.requiredImports,
      platformProviders,
      applicationProviders,
      isInternalImport,
      selectProviderKind);
    if (providerModules.size !== requiredModules.size) {
      throw new TypeError("every required import module must have one selected provider");
    }
    return Object.freeze({
      manifest,
      request,
      platformProviders: platformProviders.values,
      applicationProviders: applicationProviders.values,
    });
  });
}

function validatePlatformProviders(
  values,
  requiredModules,
  providerModules,
  grants,
  selectProviderKind,
  validateProviderMetadata) {
  const providers = new Map();
  const snapshots = values.map(value => {
    const snapshot = validateProviderMetadata({ kind: "platform", provider: value });
    if (readProviderKind(selectProviderKind, snapshot.module) !== "platform") {
      throw new TypeError("platform provider module must be reserved");
    }
    validateRequiredProvider(snapshot.module, requiredModules, providerModules);
    if (!isGranted(snapshot.capability, grants)) {
      throw new TypeError(`platform capability '${snapshot.capability}' was not granted`);
    }
    providers.set(snapshot.module, snapshot);
    return snapshot;
  });
  return Object.freeze({ byModule: providers, values: Object.freeze(snapshots) });
}

function validateApplicationProviders(
  values,
  requiredModules,
  providerModules,
  selectProviderKind,
  validateProviderMetadata) {
  const providers = new Map();
  const snapshots = values.map(value => {
    const snapshot = validateProviderMetadata({ kind: "application", provider: value });
    if (readProviderKind(selectProviderKind, snapshot.module) !== "application") {
      throw new TypeError("application provider module must not be reserved");
    }
    validateRequiredProvider(snapshot.module, requiredModules, providerModules);
    providers.set(snapshot.module, snapshot);
    return snapshot;
  });
  return Object.freeze({ byModule: providers, values: Object.freeze(snapshots) });
}

function validateRequiredProvider(module, requiredModules, providerModules) {
  if (!requiredModules.has(module)) {
    throw new TypeError(`selected provider module '${module}' is not required`);
  }
  if (providerModules.has(module)) {
    throw new TypeError(`selected provider module '${module}' is duplicated`);
  }
  providerModules.add(module);
}

function validateApplicationBindings(manifest, request, requiredModules, applications) {
  const artifacts = new Map(manifest.artifacts.map(item => [item.relativePath, item]));
  const bindings = new Map(request.applicationImports.map(item => [item.module, item]));
  for (const binding of request.applicationImports) {
    if (!requiredModules.has(binding.module)) {
      throw new TypeError(`application import module '${binding.module}' is not required`);
    }
    if (!applications.byModule.has(binding.module)) {
      throw new TypeError(`application import module '${binding.module}' has no selected provider`);
    }
    const artifact = artifacts.get(binding.artifactPath);
    if (artifact === undefined) {
      throw new TypeError(`application import artifact '${binding.artifactPath}' is missing`);
    }
    if (artifact.role !== "application-import"
        || artifact.mediaType !== "text/javascript"
        || artifact.sha256 !== binding.sha256) {
      throw new TypeError(`application import module '${binding.module}' does not match its artifact`);
    }
  }
  for (const provider of applications.values) {
    if (!bindings.has(provider.module)) {
      throw new TypeError(`application provider module '${provider.module}' has no request binding`);
    }
  }
}

function validateRequiredImports(
  requiredImports,
  platforms,
  applications,
  isInternalImport,
  selectProviderKind) {
  for (const required of requiredImports) {
    if (readInternalImport(isInternalImport, required.interface)) {
      continue;
    }
    const providers = readProviderKind(selectProviderKind, required.interface) === "platform"
      ? platforms
      : applications;
    const provider = providers.byModule.get(required.interface);
    if (provider === undefined) {
      throw new TypeError(`required import module '${required.interface}' has no selected provider`);
    }
    const provided = provider.functions.find(item => item.name === required.name);
    if (provided === undefined) {
      throw new TypeError(`required import '${required.interface}.${required.name}' is missing`);
    }
    if (!sameStrings(provided.parameters, required.parameters)
        || !sameStrings(provided.results, required.results)) {
      throw new TypeError(`required import '${required.interface}.${required.name}' has an incompatible signature`);
    }
  }
}

function readInternalImport(isInternalImport, module) {
  const internal = isInternalImport({ module });
  if (typeof internal !== "boolean") {
    throw new TypeError("internal-import policy returned an unsupported result");
  }
  return internal;
}

function isGranted(capability, grants) {
  switch (capability) {
    case "baseline":
    case "environment":
    case "preopenedDirectories":
      return true;
    case "network": return grants.network === "allowAll";
    case "wallClock": return grants.clocks.includes("wall");
    case "monotonicClock": return grants.clocks.includes("monotonic");
    case "randomness": return grants.randomness;
  }
}

function sameStrings(left, right) {
  return left.length === right.length && left.every((value, index) => value === right[index]);
}

function readProviderKind(selectProviderKind, module) {
  const kind = selectProviderKind({ module });
  if (kind !== "platform" && kind !== "application") {
    throw new TypeError("provider-kind selector returned an unsupported kind");
  }
  return kind;
}

function assertExactDataObject(value, keys, label) {
  if (value === null || typeof value !== "object" || Array.isArray(value)
      || Object.getOwnPropertySymbols(value).length !== 0) {
    throw new TypeError(`${label} is invalid`);
  }
  const prototype = Object.getPrototypeOf(value);
  const descriptors = Object.getOwnPropertyDescriptors(value);
  const actualKeys = Object.keys(descriptors).sort();
  if (prototype !== null && prototype !== Object.prototype
      || actualKeys.length !== keys.length
      || actualKeys.some((key, index) => key !== keys[index])
      || Object.values(descriptors).some(descriptor => !descriptor.enumerable || !("value" in descriptor))) {
    throw new TypeError(`${label} shape is invalid`);
  }
}
